### PR TITLE
ReactNative Remove circular type reference from NamedStyled

### PR DESCRIPTION
**Pull request by zhigang1992**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://github.com/Microsoft/TypeScript/issues/27421, https://github.com/Microsoft/TypeScript/issues/27707, https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29651, #29265>

- [ ] Increase the version number in the header if appropriate.

Should we increment the version number for this? If use was using `NameStyled` in their code then this will be a breaking change.

### CODE REVIEW

File: types/react-native/test/index.tsx [hunk 1]
```diff
@@ -112,7 +112,7 @@ BackHandler.addEventListener("hardwareBackPress", () => {}).remove();
 
 BackAndroid.addEventListener("hardwareBackPress", () => {});
 
-interface LocalStyles {
+type LocalStyles = {
```
Comment (zhigang1992) [inline types/react-native/test/index.tsx]: I'm not 100% sure about this change but without it. it'll complaint about `LocalStyles` can not do string index?

Comment (sandersn) [reply types/react-native/test/index.tsx]: Yeah, the compiler assumes that `interface LocalStyles` might be subclassed and of course `interface SubLocalStyle extends LocalStyles` has no constraint that requires its properties to be of any particular type.

`type LocalStyles`, on the other hand, can't be subclassed, so it's fine to check its properties for assignability to a string index.

**Review by sandersn — CHANGES_REQUESTED** [review]
This cleanup is nice, but if it breaks existing code (like the tests), then it might be a bad idea.

**Comment by zhigang1992** [comment]
IMHO, I think it might actually make sense to constrain the method to a key of strings.

You can't / shouldn't use anything that is not a string with `create` method from `StyleSheet`, and if you do, the runtime will throw anyway.

**Comment by vovkasm** [comment]
This was closed, but seems that fix is correct! According to disscussion here: https://github.com/Microsoft/TypeScript/issues/27421#issuecomment-431174748

**Comment by zhigang1992** [comment]
@vogievetsky yeah, but like @sandersn pointed out, it's breaking the existing change. And I have no idea how to process furthur. 😭 

Perhaps we should increment the version and put this through?

**Comment by vovkasm** [comment]
react-native typings previously broke existing code many times with small changes... one time more? :-)

Moreover fix for code is as simple as removing template parameter, isn't it?

Maintainers, what do you think? Current situation is so that without this change, typings didn't work for typescript >=3.1

**Comment by sunnylqm** [comment]
@zhigang1992 @vovkasm There happens to be a [`typesVersions `](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html) introduced in ts 3.1  to distinguish versions.

**Comment by vovkasm** [comment]
Yea, but how typesVersions can help here?  This is logical bug fix not exactly for ts >=3.1, but for all ts versions.

**Comment by sunnylqm** [comment]
I don't get your point. 
Since current version works on ts 3.0 (no matter whether it works in a "proper" way), but not on ts 3.1. so I assume that this fix will break on ts 3.0?
So I mean to apply this fix for ts >=3.1, and keep current version/files for ts 3.0 so that (probably?) won't break anything.

**Comment by vovkasm** [comment]
> so I assume that this fix will break on ts 3.0?

This fix will not break on ts 3.0. It is BREAKING change, only because users that imports and use `NamedStyles` type will need to do replace in their code `NamedStyles<string>` -> `NamedStyles`, because fix removes template argument from type `NamedStyles`.

And from other side. If you apply this fix only for ts >= 3.1, users code that use `NamedStyles` still will break after upgrade.

So I don't understand how typesVersions can help here.

**Comment by sunnylqm** [comment]
|        | Not merge | Merge without typesVersions | Merge with typesVersions |   |
|--------|-----------|-----------------------------|--------------------------|---|
| ts 3.0 | ok        | break(need user changes)    | ok                       |   |
| ts 3.1 | break     | need user changes           | need user changes        |   |
|        |           |                             |                          |   |

**Pull request closed without merging**